File: vice-running-in-cluster/handler.go
package function

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"

	"github.com/dghubble/sling"
	"github.com/pkg/errors"
	log "github.com/sirupsen/logrus"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
)

// Analyses contains the list of analyses running the cluster.
type Analyses struct {
	Analyses []string `json:"analyses"`
}

// IDRequest is the format for requests to the get-analysis-id function
type IDRequest struct {
	ExternalID string `json:"external_id"`
}

// IDResponse is the format for response bodies from the get-analysis-id funciton
type IDResponse struct {
	ID string `json:"id"`
}

// Handle is the entrypoint for OpenFaaS
func Handle(w http.ResponseWriter, r *http.Request) {
	log.SetReportCaller(true)

	viceNamespace := os.Getenv("VICE_NAMESPACE")
	gatewayURL := os.Getenv("GATEWAY")

	if viceNamespace == "" {
		log.Fatal("VICE_NAMESPACE must be set")
	}

	if gatewayURL == "" {
		log.Fatal("GATEWAY must be set")
	}

	gateway := sling.New().Base(gatewayURL)

	config, err := rest.InClusterConfig()
	if err != nil {
		log.Fatal(errors.Wrapf(err, "error loading the config inside the cluster"))
	}

	clientset, err := kubernetes.NewForConfig(config)
	if err != nil {
		log.Fatal(errors.Wrap(err, "error creating clientset from config"))
	}

	output := &Analyses{
		Analyses: []string{},
	}

	deplist, err := clientset.AppsV1().Deployments(viceNamespace).List(metav1.ListOptions{})
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	for _, deployment := range deplist.Items {
		IDReq := &IDRequest{
			ExternalID: deployment.Name,
		}
		IDResp := &IDResponse{}
		resp, err := gateway.Path("/function/get-analysis-id").BodyJSON(IDReq).ReceiveSuccess(IDResp)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		if resp.StatusCode < 200 || resp.StatusCode > 399 {
			http.Error(w, fmt.Errorf("status code from get-analysis-id was %d", resp.StatusCode).Error(), http.StatusInternalServerError)
			return
		}

		output.Analyses = append(output.Analyses, IDResp.ID)
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(output)
}

File: vice-running-in-cluster/handler.py
import json
import os
import logging

import requests
from kubernetes import client, config

has_loaded_k8s_config = False

vice_namespace = "vice-apps"

def init_k8s_config():
    global has_loaded_k8s_config

    if not has_loaded_k8s_config:
        config.load_incluster_config()
        has_loaded_k8s_config = True

def get_analysis_id(external_id):
    global vice_namespace
    
    graphql_url = os.environ["GRAPHQL_URL"]

    response = requests.post(
        graphql_url,
        json={
            'query' : """
                query GetAnalysisID($externalID: String!) {
                    analysis: analysisByExternalID(externalID: $externalID) {
                        id
                    }
                }
            """,
            'variables' : {
                'externalID' : external_id
            }
        }
    )

    response.raise_for_status()

    logging.warning("response: %s" % response.text)

    parsed_response = response.json()

    return parsed_response["data"]["analysis"]["id"]

def handle(req):
    init_k8s_config()

    api = client.AppsV1Api()

    response = api.list_namespaced_deployment(vice_namespace)
    external_ids = list(map(lambda x : x.metadata.name, response.items))
    analysis_ids = list(map(get_analysis_id, external_ids))

    return json.dumps({
        "analyses" : analysis_ids
    })
